Add registration API for torch.compile-eager

EikanWang opened the pull request:
This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125831
* #125819
* #125308
* #124926
* #124070
* #124177
* #116368
* __->__ #121387



cc @voznesenskym @penguinwu @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D57164385](https://our.internmc.facebook.com/intern/diff/D57164385)

EikanWang committed:
Add registration API for torch.compile-eager

[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we are trying to provide a registration mode to implement a single aten operation on the top of `torch.compile` and then register to aten. 




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -105,91 +105,86 @@ struct EnableHermeticPyObject {
-class PythonKernelHolder : public c10::OperatorKernel {
-  c10::SafePyObject func_;
-  c10::DispatchKey dispatch_key_;
+PythonKernelHolder::PythonKernelHolder(
```
We move the `PythonKernelHolder` declaration from the source file to the header file. W/ this change, we have to refine the `PythonKernelHolder` a little bit. From the implementation perspective, we do NOT change any implementation logic but just the code style and the declaration and definition per class member function.

EikanWang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -345,6 +340,33 @@ void initDispatchBindings(PyObject* module) {
           py::arg("dispatch") = "",
           py::arg("debug") = "impl_t_t")
+      .def(
+          "impl_t_c",
```
We define a registration API dedicated to `eager-through-torch.compile` in case we need to pass additional information to load cache or other optimizations. Compared to `imply,` there is no big difference in the implementation now. But it serves for different purpose.

ezyang replied:
The name here seems cargo culted from the testing impl names, e.g., `impl_t_t`. Why not just give this a longer, more descriptive name. As is, I have no idea what this function does.

ezyang replied:
t_c is the wrong name. t_t meant "takes one tensor, returns one tensor". You have no such constraint here, and what does returning "c" even mean.

ezyang replied:
I think this depends on how the rest of the cache hit API looks. But I would hope to see something that mentions AOTI wrapper, and also that this is the Python fallback when cache miss.

ezyang replied:
I think you also need a protocol for filling the cache on miss

EikanWang commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,55 @@
+  // TODO(Eikan): By now, we always fallback to python python_kernel_holder_ to
+  // simulate the cache miss behavior. We will add cache lookup later
+  // when the design is mature.
+  python_kernel_holder_(op, keyset, stack);
```
We utilize the `PythonKernelHolder` as the fallback path for cache miss. For a cache hit, we will load its corresponding kernel and run it directly. For a cache miss, we need to call the registered python kernel through `PythonKernelHolder`.

jansel replied:
Is this code unreachable in this PR because `cache_lookup` just returns false?

I would expect this to call into the C++ kernel code and not go through Python.

EikanWang commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,55 @@
+  } else if (dispatch_key_ == c10::DispatchKey::XPU) {
+    device_opt_ = c10::Device(c10::DeviceType::XPU, 0);
+  } else {
+    device_opt_ = c10::Device(c10::DeviceType::COMPILE_TIME_MAX_DEVICE_TYPES);
```
Good idea. I will refine it.

EikanWang replied:
@desertfire , I refined the code a little bit by replacing the `if-else` block with `c10::dispatchKeyToDeviceType`

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

desertfire commented on test/inductor/test_torchinductor.py:
```diff
@@ -718,6 +719,74 @@ def fn(a, b):
+    def test_torch_compile_override_registration(self):
+        dynamic = True
+        namespace_name = "aten"
+        dispatch_key = "CPU"
```
Can we test CUDA as well?

EikanWang replied:
Sure. I will refine the test case a little bit.

desertfire reviewed (approved):
LGTM. cc @ezyang  in case he has any comment to the changes to the dispatcher part.

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

ezyang commented on torch/library.py:
```diff
@@ -110,6 +110,28 @@ def define(self, schema, alias_analysis="", *, tags=()):
         _defs.add(qualname)
         return result
 
+    def _impl_t_c(self, op_name, fn, dispatch_key=''):
```
docs???

zou3519 replied:
Please change the name to something more readable. No one is going to know what `_impl_t_c` is (I still don't know what it is)

ezyang commented on torch/csrc/inductor/aoti_eager/kernel_holder.h:
```diff
@@ -0,0 +1,39 @@
+
+namespace torch::inductor {
+
+class AOTIPythonKernelHolder : public c10::OperatorKernel {
```
docs?

ezyang commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,47 @@
+
+void AOTIPythonKernelHolder::canonicalizeOpName() {
+  // Canonicalize the op_name as a valid directory name
+  std::replace(op_name_.begin(), op_name_.end(), '.', '_');
```
This mangling is insufficient to guarantee uniqueness. cc @zou3519

Also, dots are valid in directory names, what's the problem

ezyang commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,47 @@
+  if (start_pos != std::string::npos) {
+    op_name_.replace(start_pos, to_remove.length(), "");
+  }
+}
```
This entire function seems very ill considered

EikanWang replied:
The pattern of `op_name_` is `aten::${op_name}.${schema_overload_name}`. This function transfers the pattern to `${op_name}_${schema_overload_name}` for cache loading. May I know your detailed concerns? @ezyang

ezyang replied:
foo.bar_baz will be ambiguous with foo_bar.baz. And I'm just generally suspicious of this amount of string mangling in C++ code

ezyang replied:
You're also... removing the namespace? By string search replacing? It just all seems suspicious!

zou3519 replied:
+1 this is pretty suspicious. Your mangling is not unique, and you're dropping the namespace. 

Ed, I'm not sure want to use the mangling scheme I'm developing (in a separate PR) here (we certainly can, but it's meant to mangle Python FQNs, not dispatcher op names). We would need to extend it a little.

EikanWang replied:
Thanks for the elaboration.

The full path will be `${TORCHINDUCTOR_CACHE_DIR}/aten_eager/${canonicalized_op_name}.json`. (Sorry for that I only added comments in another PR).

We intend to only support aten operations through `torch.compile`. Based on this intension, I suppose the namespace is always `aten` and its parent folder has contained `aten` literal, therefore, I remove it. And I can add it back if you have concerns. 

The key problem is if the name is unique. I statistic all schemas of `native_functions.yaml` and `RegistrationDeclarations.h`. There are 3103 in total. And almost all the schemas are unique except `_test_autograd_multiple_`. It does not have namespace. Since we only intend to support aten operations, I suppose we do not need to take `_test_autograd_multiple_` into account.

zou3519 replied:
@EikanWang we want this infrastructure to be robust against some future use cases:
- if someone adds an aten::foo.bar_baz op and an aten::foo_bar.baz op in the future, then this will break.
- I'm not convinced that we just want to do this for aten ops. There are a lot of built-in operators in PyTorch that are not in the aten namespace (e.g. prim::, inductor::, quantized::, etc).

So, we should add the namespace back and develop a robust mangling scheme

EikanWang replied:
Thanks. Yes, sure. I will add the namespace back and keep dots.

EikanWang replied:
@zou3519 , I refined the PR. W/ the new design, the namespace will be part to cache directory - https://github.com/pytorch/pytorch/pull/116368/files#diff-8dfc175f817a04eb10c183509564129a91f32db31190e7b4ad3fd9653337c6f6R1528.

zou3519 reviewed (changes requested):
Shout when you've updated this. To summarize, the concerns are:
1) docs
2) filenames
3) naming of the new APIs

EikanWang commented:
> Shout when you've updated this. To summarize, the concerns are:
> 
> 1. docs
> 2. filenames
> 3. naming of the new APIs

I will sync with Bin to figure out some implementation details and keep you posted.

EikanWang commented:
@zou3519 , I synced with Bin and I will organize the ghstack a little bit by stacking this PR on the top of AOTI-based kernel implementation and I'm working on it. I will keep you posted as long as the PR is ready.

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+    const c10::IValue& ivalue,
+    const c10::Device& device,
+    std::vector<at::Tensor>& inputs);
+std::unordered_map<IValueType, HandlerFunc> handlers_ = {
+    {IValueType::Tensor, &HandleTensor},
+    {IValueType::TensorList, &HandleTensorList},
+    {IValueType::OptionalTensorList, &HandleOptionalTensorList},
+    {IValueType::Scalar, &HandleScalar}};
```
Consider to cover these scenarios in UTs?

EikanWang replied:
Regarding the `scalar` support, I will add another PR to support scalar. For the other options, I will add more test cases.

EikanWang commented on torch/library.py:
```diff
@@ -123,6 +123,47 @@ def define(self, schema, alias_analysis="", *, tags=()):
         return result
 
+    def _impl_with_aoti_compile(self, op_name, op_overload_name, dispatch_key='', fallback_fn=None):
+        r'''Registers the AOT-compiled implementation for the operator defined in the library.
```
Sounds good. I will update it.

EikanWang commented on torch/library.py:
```diff
@@ -123,6 +123,47 @@ def define(self, schema, alias_analysis="", *, tags=()):
+        r'''Registers the AOT-compiled implementation for the operator defined in the library.
+
+        Args:
+            op_name: operator name
+            op_overload_name: operator overload name
```
Sounds good. Yes, we should.

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the registered python kernel

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR.




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang commented:
We refined the PR by adding namesapce as part of cache directory and keeping dots. The cache directory is 
```python
device_kernel_cache = Path.home() / ".cache" / "torch" / "aoti_eager" / name_space.lower() / device_type.lower()
```

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jansel commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+    const c10::IValue& ivalue,
+    const c10::Device& device,
+    std::vector<at::Tensor>& inputs) {
+  return HandleTensorList(ivalue, device, inputs);
```
Don't we need to handle the optional not provided case?

jansel commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+    const c10::Device& device,
+    std::vector<at::Tensor>& inputs) {
+  IValueType ivalue_type = IValueType::Invalid;
+  if (ivalue.isTensor()) {
+    ivalue_type = IValueType::Tensor;
+  } else if (ivalue.isTensorList()) {
+    ivalue_type = IValueType::TensorList;
+  } else if (ivalue.isOptionalTensorList()) {
+    ivalue_type = IValueType::OptionalTensorList;
+  } else if (ivalue.isScalar()) {
+    ivalue_type = IValueType::Scalar;
+  }
```
Why not just call the handler functions directly here?  That would be less code and faster since it avoids an unorered_map lookup.

jansel commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+  }
+
+  std::vector<at::Tensor> inputs;
+  if (unpackTensors(*stack, device, inputs)) {
```
What happens if this returns false?  It seems like we just exit the function.

I was expecting an assert.

jansel commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+  Invalid,
+};
+
+static bool HandleTensor(
```
Doesn't most of the codebase use lowercase function names?  We should match the style.

jansel commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+  return false;
+}
+
+bool unpackTensors(
```
Use consistent coding style.  Uppercase above.  Camelcase here.  Underscores below.

jansel commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+      has_fall_back_(func.ptr() != Py_None),
+      device_opt_(c10::Device(c10::dispatchKeyToDeviceType(dispatch_key_), 0)),
+      pyinterpreter_(getPyInterpreter()) {
+  (void)is_symbolic_; // Suppress unused variable warning
```
Why do we need an unused arg here?

jansel commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+
+  auto kernel_lib_path = produce_aoti_kernel_lib(op, keyset, stack);
+  if (kernel_lib_path.empty()) {
+    TORCH_CHECK(
+        has_fall_back_,
+        "Failed to produce kernel libarary by using AOTI and no fall back");
+    python_kernel_holder_(op, keyset, stack);
+    return;
```
Are we hitting this in practice?  Should this just be an error?

EikanWang replied:
The registration API(`_impl_with_aoti_compile`) provides a fallback function input in case AOTI could not produce kernel. For example, users could provide a numpy implementation for a particular aten operation if AOTI failed to produce kernel. Do you think we do not need to provide such a fallback function? @jansel , @desertfire

jansel commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+    const c10::Device& device,
+    std::vector<at::Tensor>& inputs) {
+  for (const auto& item : ivalue.toListRef()) {
+    if (!item.isNone()) {
```
Is ignoring none correct here?  Doesn't that change the args?

EikanWang replied:
I refined the PR. Regarding the `OptionalTensorList`, we will check if the `std::optional<at::Tensor>` has value.

jansel commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+  });
+
+  if (op_py_func.ptr() == nullptr || op_py_func.ptr() == Py_None) {
+    TORCH_WARN(
+        "Failed to get python operation. Operator Name is ",
+        op.operator_name().name,
+        ", Overload Name is ",
+        overload_name);
```
Should this be an error, when does it happen?

EikanWang replied:
Thanks. I replaced it with assert.

jansel commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+      py::module::import("torch._export").attr("aot_compile");
+  if (aot_compile_function.ptr() == nullptr ||
+      aot_compile_function.ptr() == Py_None) {
+    TORCH_WARN("Failed to import - torch._export.aot_compile");
+    return kernel_lib_path;
```
Should this be an error, when does it happen?

EikanWang replied:
Thanks. I replaced it with assert.

jansel commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+  if (result.ptr() != nullptr && result.ptr() != Py_None) {
+    kernel_lib_path = py::cast<std::string>(result);
+    if (kernel_lib_path.empty()) {
+      TORCH_WARN(
+          "Kernel library is not generated by AOTI for ",
+          c10::DeviceTypeName(device_opt_.value().type()),
+          ". Operator Name is ",
+          op.operator_name().name,
+          ", Overload Name is ",
+          overload_name);
```
Should this be an error, when does it happen?

jansel commented on torch/csrc/inductor/aoti_eager/kernel_holder.cpp:
```diff
@@ -0,0 +1,283 @@
+          overload_name);
+    }
+  } else {
+    TORCH_WARN(
+        "AOTI kernel library is not generated for ",
+        c10::DeviceTypeName(device_opt_.value().type()),
+        ". Operator Name is ",
+        op.operator_name().name,
+        ", Overload Name is ",
+        overload_name);
```
Should this be an error, when does it happen?

EikanWang replied:
I will replace it with assert because we will guarantee the aoti compile always returns a value, even in the event of errors.

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang reviewed (commented):
@jansel, @desertfire  , do you think we should provide the fallback function for `impl_with_aoti_compile` if AOTI could not produce kernel library rather than just throw exception? It will impact how to handle the error.

https://github.com/pytorch/pytorch/pull/121387/files#diff-f6b9e73c52d37bbfcc9289a0af9f8f5d2a28c3040fcd4cc07cdcf32ab96baa4aR142

zou3519 commented on torch/library.py:
```diff
@@ -139,6 +139,58 @@ def _register_fake(self, op_name, fn, _stacklevel=1):
         handle = entry.abstract_impl.register(func_to_register, source)
         self._registration_handles.append(handle)
 
+    def _impl_with_aoti_compile(self, op_name, op_overload_name='', dispatch_key='', fallback_fn=None):
```
To match `impl`, we shouldn't have an op_overload_name -- all the info is included in op_name.

zou3519 commented on torch/library.py:
```diff
@@ -139,6 +139,58 @@ def _register_fake(self, op_name, fn, _stacklevel=1):
+            op_overload_name: operator overload name
+            dispatch_key: dispatch key that the input function should be registered for. By default, it uses
+                          the dispatch key that the library was created with.
+            fallback_fn: This is a user-defined function that is used as a backup plan in case the AOT inductor
+                         fails to generate a kernel. The purpose of this function is to ensure that the operation
+                         can still be executed even if the kernel production fails. If the AOT inductor fails and
+                         no fallback function is provided, an error will be raised during the operation.
```
How often do we actually need a custom fallback? It might be better to implement this at the Dispatcher fallback level. For example, if we're trying to do this for XPU (this is the use case, right?), then the XPU dispatch key can have a fallback of moving the tensors onto CPU, calling the torch operation, and then moving the tensor back onto XPU.

EikanWang replied:
Thanks, @zou3519 .

> How often do we actually need a custom fallback?

Eventually, we rarely need the fallback as AOTI can support ATen operations well.

> It might be better to implement this at the Dispatcher fallback level. ... have a fallback of moving the tensors onto CPU, calling the torch operation, and then moving the tensor back onto XPU.

Regarding the implementation at the Dispatcher level, it may not be triggered because the dispatch key has been registered with AOTI kernel holder. But, we can still fallback to CPU implement to serve cache miss. Thanks for the proposal.

In conclusion, we should remove `fallback_fn`. @zou3519 , are we on the same page?

zou3519 replied:
> Regarding the implementation at the Dispatcher level, it may not be triggered because the dispatch key has been registered with AOTI kernel holder. But, we can still fallback to CPU implement to serve cache miss. Thanks for the proposal.

If AOTI fails to compile a kernel, then we probably don't want to actually register the kernel to the XPU dispatch key. This would let us get to the XPU dispatch key's fallback.

> In conclusion, we should remove fallback_fn, right? @zou3519 , are we on the same page?

Yeah, it sounds like we don't need it right now

EikanWang commented on torch/library.py:
```diff
@@ -139,6 +139,58 @@ def _register_fake(self, op_name, fn, _stacklevel=1):
+        Args:
+            op_name: operator name or OpOverload object.
+            op_overload_name: operator overload name
+            dispatch_key: dispatch key that the input function should be registered for. By default, it uses
+                          the dispatch key that the library was created with.
```
Thanks, I will add such check.

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang commented:
@zou3519 , may I know if I have addressed your comments?

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented on torch/library.py:
```diff
@@ -154,6 +154,48 @@ def _register_fake(self, op_name, fn, _stacklevel=1):
+            >>> my_lib = Library("aten", "IMPL")
+            >>> my_lib._impl_with_aoti_compile("div.Tensor", "CPU")
+        '''
+        if dispatch_key == '':
+            dispatch_key = self.dispatch_key
+        assert dispatch_key in ['CPU', 'CUDA', 'XPU']
```
our check should probably just be that the dispatch key contains DispatchKey::Dense (so that it can support all devices, not just these three)

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang commented:
@pytorchbot merge

EikanWang commented:
@pytorchbot merge

kit1980 commented:
I need to revert this as it's failing internally 

```
ERROR: test_torch_compile_override_registration_cuda (caffe2.test.inductor.test_torchinductor.GPUTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/dev/shm/uid-30083/32adf917-seed-nspid4026541534_cgpid12437421-ns-4026541785/caffe2/test/inductor/test_torchinductor.py", line 9819, in new_test
    return value(self)
  File "/dev/shm/uid-30083/32adf917-seed-nspid4026541534_cgpid12437421-ns-4026541785/caffe2/test/inductor/test_torchinductor.py", line 808, in test_torch_compile_override_registration
    res_array.append(getattr(torch, unary_op_name)(x))
RuntimeError: Error in dlopen: /re_tmp/tmpxk4fyq1k/cw74bdbcowopv7pivuo7czemlte5oieduln75gcpslbqehcdy7q2/cstf4iodqtmpabtysxkoezwgtt5d5nv6ooe6q4vaztopatoszyl7.so: undefined symbol: aoti_torch_device_type_cpu
```

I think the new file needs to be added to the buck dependencies.
@zou3519 @jansel see D56736862 if you want to help to re-land this.

kit1980 commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst

EikanWang commented:
@desertfire , may I know if it means the internal CI only support CUDA?

EikanWang commented:
> test_torch_compile_override_registration_cuda (caffe2.test.inductor.test_torchinductor.GPUTests)

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

EikanWang committed:
Update on "Add registration API for torch.compile-eager"

This PR is a follow-up of RFC https://github.com/pytorch/pytorch/issues/115545.

In this PR, we intend to provide a registration API dedicated to eager-through-torch.compile. The major workflow of this API will be as follows.

- Load cache
- Check cache according to the input tensors
  - Cache Hit: Run the cached kernel directly
  - Cache Miss: Run the AOTI to produce kernel and run the produced kernel. If AOTI fails to produce the kernel, invoke the python fallback function.

Currently, this PR always fallback to python kernel now and cache mechanism will be implemented in another PR - https://github.com/pytorch/pytorch/pull/116368




cc voznesenskym penguinwu jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

atalman commented:
@atalman has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57164385).

atalman commented:
@pytorchmergebot merge -f "Already landed in fbcode"
